design/one-pager-crossplane-state-metrics.md

```diff
@@ -0,0 +1,640 @@
+
+### Architecture
+
+The operator is deployed as an independent controller with its own RBAC, CRDs, and reconciliation loops.
+
+**Core Components:**
+
+1. **ClusterMetric Controller**
+   - Watches ClusterMetric CRDs for configuration changes
+   - Discovers and monitors cluster-scoped target resources (Providers, cluster XRs, etc.)
+   - Extracts values and labels using JSONPath expressions
+   - Registers Prometheus metrics based on ClusterMetric definitions
+   - Tracks state transitions for counter metrics
+
+2. **Metric Controller**
+   - Watches Metric CRDs (namespace-scoped) for configuration changes
+   - Discovers and monitors namespace-scoped target resources (XRs, Claims, etc.)
+   - Same extraction and metric generation logic as ClusterMetric
+   - Provides namespace-level isolation for metrics
+
+3. **Prometheus Metrics Endpoint**
+   - Exposes metrics at `/metrics` endpoint
+   - Standard Prometheus scrape target
+   - Serves both gauge (current state) and counter (state transitions) metrics
+   - All metrics follow pattern: `crossplane_{metadata.name}` and `crossplane_{metadata.name}_total`
+
+### API Design
+
+#### Metric and ClusterMetric
+
+The operator provides two CRD types for defining metrics:
+
+- **Metric**: Namespace-scoped, for monitoring resources within a specific namespace
+- **ClusterMetric**: Cluster-scoped, for monitoring cluster-wide resources
+
+Both share the same core spec structure:
+
+```yaml
+apiVersion: metrics.crossplane.io/v1alpha1
+kind: ClusterMetric
+metadata:
+  name: xnetwork-ready-status
+spec:
+  # Help text for the metric (Prometheus HELP)
+  help: "XNetwork ready status (1=ready, 0=not ready)"
+
+  # Reconciliation interval (default: 10m)
+  interval: 1m
+
+  # Target resource to monitor
+  target:
+    group: aws.platform.upbound.io
+    kind: XNetwork        # Supports wildcard "*" for all kinds
+    version: v1alpha1
+
+  # Optional: Filter resources
+  labelSelector: "team=platform"
+  fieldSelector: "metadata.namespace=production"
+
+  # Define metric values to extract (using JSONPath)
+  metrics:
+  - name: ready_status
+    path: "status.conditions[?(@.type=='Ready')].status"
+
+  # Define label dimensions (using JSONPath)
+  labels:
+  - name: resource_name
+    path: "metadata.name"
+  - name: region
+    path: "spec.parameters.region"
+  - name: team
+    path: "metadata.labels['team']"
+```
+
+**Key API Features:**
+
+1. **JSONPath Support**: Extract any field from the target resource using JSONPath expressions
+2. **Wildcard Kind Monitoring**: Use `kind: "*"` to monitor all resource types in a group/version
+3. **Flexible Filtering**: Label and field selectors to target specific resources
+4. **Configurable Intervals**: Per-metric reconciliation intervals to balance freshness vs. load
```

Comment:
Pretty sure it was already your intention, but a configurable global default value through a flag would also be nice.

Reply:
sure an global like our core crossplane have one